pkg/app/piped/imagewatcher/watcher.go

```diff
@@ -13,68 +13,179 @@
 	return &watcher{
-		timer:  time.NewTimer(interval),
-		logger: logger,
+		config:    cfg,
+		gitClient: gitClient,
+		logger:    logger.Named("image-watcher"),
 	}
 }
 
+// Run spawns goroutines for each image provider.
 func (w *watcher) Run(ctx context.Context) error {
+	// Pre-clone to cache the registered git repositories.
+	for _, r := range w.config.Repositories {
+		repo, err := w.gitClient.Clone(ctx, r.RepoID, r.Remote, r.Branch, "")
+		if err != nil {
+			w.logger.Error("failed to clone repository",
+				zap.String("repo-id", r.RepoID),
+				zap.Error(err),
+			)
+			return err
+		}
+		w.gitRepos.Store(r.RepoID, repo)
+	}
+
+	for _, cfg := range w.config.ImageProviders {
+		p, err := imageprovider.NewProvider(&cfg, w.logger)
+		if err != nil {
+			return err
+		}
+
+		go w.run(ctx, p, cfg.PullInterval.Duration())
+	}
+	return nil
+}
+
+// run periodically compares the image stored in the given provider and one stored in git.
+// And then pushes those with differences.
+func (w *watcher) run(ctx context.Context, provider imageprovider.Provider, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
 	for {
 		select {
 		case <-ctx.Done():
-			return nil
-		case <-w.timer.C:
-			reposInReg := w.fetchFromRegistry()
-			reposInGit := w.fetchFromGit()
-			outdated := calculateChanges(reposInReg, reposInGit)
-			if err := w.update(outdated); err != nil {
+			return
+		case <-ticker.C:
+			targets := w.collectTargets(ctx, provider)
+			outdated, err := determineUpdates(ctx, targets, provider)
+			if err != nil {
+				w.logger.Error("failed to determine which one should be updated", zap.Error(err))
+				continue
+			}
+			if len(outdated) == 0 {
+				w.logger.Info("no image to be updated")
+				continue
+			}
+			if err := update(outdated); err != nil {
 				w.logger.Error("failed to update image", zap.Error(err))
+				continue
 			}
 		}
 	}
-	return nil
 }
 
-func (w *watcher) fetchFromRegistry() imageRepos {
-	return nil
+// collectTarget collects target images for each git repository.
+func (w *watcher) collectTargets(ctx context.Context, provider imageprovider.Provider) (targets []config.ImageWatcherTarget) {
+	w.gitRepos.Range(func(key, value interface{}) bool {
+		id, ok := key.(string)
+		if !ok {
+			w.logger.Error("unknown key type found")
+			return true
+		}
+		repo, ok := value.(git.Repo)
+		if !ok {
+			w.logger.Error("unknown repo type found")
+			return true
+		}
+		branch := repo.GetClonedBranch()
+		if err := repo.Pull(ctx, branch); err != nil {
+			w.logger.Error("failed to update repository branch",
+				zap.String("repo-id", id),
+				zap.Error(err),
+			)
+			return true
+		}
+
+		includes := make([]string, 0)
+		excludes := make([]string, 0)
+		for _, target := range w.config.ImageWatcher.Targets {
+			if target.RepoID != id {
+				continue
+			}
+			includes = append(includes, target.Includes...)
+			excludes = append(excludes, target.Excludes...)
+		}
```

Comment:
nit: Instead of finding and saving these `include/exclude` list, just finding and marking the target.

Reply:
Okay, let me address it when I work on this 🙏 :
https://github.com/pipe-cd/pipe/blob/1c4d30e9b310b4e180af234efcffe9e142ba4d7d/pkg/config/image_watcher.go#L31-L32